File: src/hebrew_anagram/matcher.py
"""Letter-availability matching for Hebrew anagram checking."""

from .letters import remove_niqqud


def can_form_word(
    letters: str,
    candidate: str,
    wildcard: str = "?",
) -> bool:
    """Return whether *candidate* can be formed from the available *letters*.

    Rules:
    - Niqqud is stripped from both *letters* and *candidate* before matching.
    - Final-letter normalization is **not** applied by default; ך and כ are
      treated as distinct letters. Pass pre-normalized strings if you need
      final-letter-agnostic matching.
    - Whitespace inside *letters* is ignored (e.g. a rack stored as "א ב ג").
    - Each character in *wildcard* (default ``"?"``) can stand in for any
      single Hebrew letter from *candidate* not covered by a real letter.
    - Inputs are never mutated.

    Args:
        letters: Pool of available letters (the "rack").
        candidate: Word to check.
        wildcard: Character(s) treated as universal substitutes.

    Returns:
        ``True`` if *candidate* can be formed, ``False`` otherwise.

    Raises:
        TypeError: If *letters* or *candidate* is not a ``str``.
    """
    if not isinstance(letters, str):
        raise TypeError(f"letters must be str, got {type(letters).__name__}")
    if not isinstance(candidate, str):
        raise TypeError(f"candidate must be str, got {type(candidate).__name__}")

    # Normalise: strip niqqud, drop whitespace from the rack
    rack_clean = remove_niqqud(letters).replace(" ", "")
    word_clean = remove_niqqud(candidate)

    # Build mutable frequency maps
    rack_counts: dict[str, int] = {}
    for ch in rack_clean:
        rack_counts[ch] = rack_counts.get(ch, 0) + 1

    wildcards_available = rack_counts.pop(wildcard, 0)

    wildcards_used = 0
    for ch in word_clean:
        if rack_counts.get(ch, 0) > 0:
            rack_counts[ch] -= 1
        elif wildcards_used < wildcards_available:
            wildcards_used += 1
        else:
            return False

    return True

File: web/src/lib/matcher.ts
// Anagram matcher.
// Final-letter normalization is OFF by default — ם and מ are distinct.

import { removeNiqqud } from "./hebrew";

export interface CanFormOptions {
  wildcard?: string;
}

/**
 * True iff *candidate* can be formed from the available *letters*.
 *
 * - Niqqud is stripped from both sides before matching.
 * - Whitespace inside *letters* is ignored (so a "rack" can be entered as
 *   spaced-out characters).
 * - Each occurrence of *wildcard* in *letters* substitutes for any single
 *   Hebrew letter in *candidate* not covered by a real letter.
 * - Final letters are NOT normalized; pre-normalize both sides to make
 *   ם and מ interchangeable.
 */
export function canFormWord(
  letters: string,
  candidate: string,
  { wildcard = "?" }: CanFormOptions = {},
): boolean {
  const rack = removeNiqqud(letters).replace(/\s+/g, "");
  const word = removeNiqqud(candidate);

  const counts = new Map<string, number>();
  for (const ch of rack) counts.set(ch, (counts.get(ch) ?? 0) + 1);

  const wildcardsAvailable = counts.get(wildcard) ?? 0;
  counts.delete(wildcard);

  let wildcardsUsed = 0;
  for (const ch of word) {
    const have = counts.get(ch) ?? 0;
    if (have > 0) {
      counts.set(ch, have - 1);
    } else if (wildcardsUsed < wildcardsAvailable) {
      wildcardsUsed += 1;
    } else {
      return false;
    }
  }
  return true;
}
